### ISSUE TITLE
TensorBoardLogger not working as expected with accumulate_grad_batches>1

### ISSUE THREAD

**Issue by jopo666** [PyTorchLightning/pytorch-lightning#4304]
## 🐛 Bug

When logging inside training step to TensorBoard and using `accumulate_grad_batches > 1` inside `pl.Trainer()` the behavior is not as expected.

With `accumulate_grad_batches = 1` everything looks good.
<img width="273" alt="nice" src="https://user-images.githubusercontent.com/49716607/96869625-65b62900-1478-11eb-85fd-c3c5d219c65e.png">

With `accumulate_grad_batches = 8` the values are reported on the same step.

<img width="271" alt="not_nice" src="https://user-images.githubusercontent.com/49716607/96869632-66e75600-1478-11eb-981b-399af7dd173c.png">

### To Reproduce (sorry for not using colab)

```
import os

import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split

import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger
from torchvision.datasets.mnist import MNIST
from torchvision import transforms

class LitClassifier(pl.LightningModule):
    def __init__(self, hidden_dim=128, learning_rate=1e-3):
        super().__init__()
        self.save_hyperparameters()

        self.l1 = torch.nn.Linear(28 * 28, self.hparams.hidden_dim)
        self.l2 = torch.nn.Linear(self.hparams.hidden_dim, 10)

    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = torch.relu(self.l1(x))
        x = torch.relu(self.l2(x))
        return x

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.log("train_loss",loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

def run_test(accumulate_grad_batches,batch_size,num_workers):
    dataset = MNIST('', train=True, download=True, transform=transforms.ToTensor())
    mnist_train, mnist_val = random_split(dataset, [55000, 5000])
    train_loader = DataLoader(mnist_train,batch_size)
    val_loader = DataLoader(mnist_val,batch_size)

    model = LitClassifier()

    trainer = pl.Trainer(
        logger=TensorBoardLogger(os.getcwd()',name='bug'),
        accumulate_grad_batches=accumulate_grad_batches,
        max_epochs=2
        )
    trainer.fit(model, train_loader, val_loader)

run_test(1,32)
run_test(8,32)
```

### Expected behavior

Take a mean of the values logged at same step?

### Environment

 - PyTorch Version: 1.6
 - OS: Linux
 - How you installed PyTorch: conda
 - Python version: 3.8
 - Tensorboard: 2.3.0

**Comment by marrrcin** [comment]
To add something to the issue: when using DataParallel (`accelerator="dp"`) it might cause even worse results, i.e in W&B it looks like this:
![image](https://user-images.githubusercontent.com/6958772/97011285-c2801500-1546-11eb-8362-37f2e5e1814e.png)

I'm also logging from `training_step`.

**Comment by rohitgr7** [comment]
The first one looks like an issue with the limitations of `Tensorboard` since `global_step` is updated after every accumulated step, and since x-axis is the `global_step` so for every `global_step` it's logging 8 values(num accumulated batches). I tried this on Wandb the it's looking good there.
![Screenshot from 2020-10-24 02-45-36](https://user-images.githubusercontent.com/30778939/97055069-a199dd00-15a3-11eb-9872-b9c4ebeb6348.png)

**Comment by marrrcin** [comment]
@rohitgr7 have you checked that with or without DataParallel?

**Comment by rohitgr7** [comment]
@marrrcin don't have GPUs 😅, can't check it.

**Comment by marrrcin** [comment]
So what I've found out is the following:
1. Logging with `on_step=False` and `on_epoch=True` in `__step_end` yields fine charts on the epoch level (for any kind of step).
2. Logging with `on_step=True` and `on_epoch=False` in `training_step_end` results in messed up charts that are dependant on the `accumulate_grad_batches` parameter of the `Trainer` as in the picture:
![7E1ED620-2400-4E84-B7A3-C1E6782599D3](https://user-images.githubusercontent.com/6958772/97088369-93de6900-1630-11eb-8d9b-369060a7d944.jpg)
**Each red box is contains number of points equal to `accumulate_grad_batches`.** 

It seems to me like the problem is that for the same `Trainer`'s `global_step`, there are `accumulate_grad_batches`-calls to both `train_step` and `backward` and when you log from there, the charts are invalid. IMHO when `accumulate_grad_batches` is > 1 then there should be some callback / `on_*` event just before the optimizer step happens that will have access to the losses calculated during the accumulation step.

My current workaround to this is to store every training loss's value during the same `global_step` and only call `log` when global step increments - then the chart looks normal:
![Screenshot 2020-10-24 at 19 57 10](https://user-images.githubusercontent.com/6958772/97088721-27189e00-1633-11eb-8e25-337ec66d7175.png)

It's a nasty workaround, I hope @tchaton you will be able to fix it in more robust manner.

**Comment by marrrcin** [comment]
@tchaton any update on this?

**Comment by marrrcin** [comment]
Ping @tchaton

**Comment by edenlightning** [comment]
defining the behavior only for accumulate_grad_batches,
avoid making general logging changes
